### PR TITLE
Add experimental ts-morph declaration generator

### PR DESCRIPTION
**Pull request by mitchellhamilton**

### CODE REVIEW

**Comment by ifiokjr** [comment]
@mitchellhamilton this look interesting, what are the advantages of `ts-morph`?

**Comment by mitchellhamilton** [comment]
Using it fixes some strange things with the declaration generation in Preconstruct, I haven't been able to figure out how to make Preconstruct's declaration generation completely correct because tsc's API is really confusing.(see the test case in this PR)

**Comment by Andarist** [comment]
So this basically just uses a different API for `.d.ts` generation and doesn't actually have to handle this weird issue in any special way?

File: packages/cli/src/project.ts [hunk 1]
```diff
@@ -24,7 +24,8 @@ export class Project extends Item {
     let config = this._config.___experimentalFlags_WILL_CHANGE_IN_PATCH || {};
     return {
       newEntrypoints: !!config.newEntrypoints,
-      useSourceInsteadOfGeneratingTSDeclarations: !!config.useSourceInsteadOfGeneratingTSDeclarations
+      useSourceInsteadOfGeneratingTSDeclarations: !!config.useSourceInsteadOfGeneratingTSDeclarations,
+      useTSMorphToGenerateTSDeclarations: !!config.useTSMorphToGenerateTSDeclarations,
```
Comment (Andarist) [inline packages/cli/src/project.ts:28]: is the plan to remove `useSourceInsteadOfGeneratingTSDeclarations` now then?

**Comment by mitchellhamilton** [comment]
> So this basically just uses a different API for .d.ts generation and doesn't actually have to handle this weird issue in any special way?

Correct. Under the hood, it's still using the TypeScript compiler, ts-morph is just a wrapper that sets it up correctly.

File: packages/cli/src/project.ts (hunk 1 above)
Comment (mitchellhamilton) [reply packages/cli/src/project.ts:28]: yep

**Pull request merged**